crates/sui-indexer/src/models/epoch.rs

```diff
@@ -117,50 +117,107 @@ pub struct QueryableEpochSystemState {
     pub system_state: Vec<u8>,
 }
 
-impl StartOfEpochUpdate {
+#[derive(Default)]
+pub struct EpochStartInfo {
+    pub first_checkpoint_id: u64,
+    pub first_tx_sequence_number: u64,
+    pub total_stake: u64,
+    pub storage_fund_balance: u64,
+}
+
+impl EpochStartInfo {
     pub fn new(
-        new_system_state_summary: SuiSystemStateSummary,
         first_checkpoint_id: u64,
         first_tx_sequence_number: u64,
-        event: Option<&SystemEpochInfoEvent>,
+        epoch_event_opt: Option<&SystemEpochInfoEvent>,
+    ) -> Self {
+        Self {
+            first_checkpoint_id,
+            first_tx_sequence_number,
+            total_stake: epoch_event_opt.map(|e| e.total_stake).unwrap_or_default(),
+            storage_fund_balance: epoch_event_opt
+                .map(|e| e.storage_fund_balance)
+                .unwrap_or_default(),
+        }
+    }
+}
+
+impl StartOfEpochUpdate {
+    pub fn new(
+        new_system_state_summary: SuiSystemStateSummary,
+        epoch_start_info: EpochStartInfo,
     ) -> Self {
         Self {
             epoch: new_system_state_summary.epoch as i64,
             system_state_summary_json: serde_json::to_value(new_system_state_summary.clone())
                 .unwrap(),
-            first_checkpoint_id: first_checkpoint_id as i64,
-            first_tx_sequence_number: first_tx_sequence_number as i64,
+            first_checkpoint_id: epoch_start_info.first_checkpoint_id as i64,
+            first_tx_sequence_number: epoch_start_info.first_tx_sequence_number as i64,
             epoch_start_timestamp: new_system_state_summary.epoch_start_timestamp_ms as i64,
             reference_gas_price: new_system_state_summary.reference_gas_price as i64,
             protocol_version: new_system_state_summary.protocol_version as i64,
-            // NOTE: total_stake and storage_fund_balance are about new epoch,
-            // although the event is generated at the end of the previous epoch,
-            // the event is optional b/c no such event for the first epoch.
-            total_stake: event.map(|e| e.total_stake as i64).unwrap_or(0),
-            storage_fund_balance: event.map(|e| e.storage_fund_balance as i64).unwrap_or(0),
+            total_stake: epoch_start_info.total_stake as i64,
+            storage_fund_balance: epoch_start_info.storage_fund_balance as i64,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct EpochEndInfo {
+    pub first_tx_sequence_number: u64,
```

Comment:
This field doesn't seem to belong to EpochEndInfo. I think it may be cleaner to just pass it separately into the EndOfEpochUpdate constructor. This will also make the Default natural.